### PR TITLE
Added option to not to create empty archive

### PR DESCRIPTION
**Pull request by sharikovvladislav**
Some days ago I created this issue: https://github.com/gruntjs/grunt-contrib-compress/issues/185

I realized, that `compress` task from `grunt-contrib-compress` npm module will create empty `zip` file if files, which are specified at `files` option are not exists.

I created this pull request. I added option, which will add this missing (as I think) feature.

I have 2 questions also: 
1) Do you think this name of option is good? Or should I rename it to some another name?
2) About tests. Do you think I created correct tests for the changes?

Please, correct my PR or add some comments so I can correct the code or PR.

### CODE REVIEW

**Comment by sharikovvladislav** [comment]
I got "AppVeyor build failed" message. As I see in log this happened because system couldn't connect to GitHub on 443 port. Looks like I have to rerun this build. I can't find how I can do this.

**Comment by sharikovvladislav** [comment]
Hm, I updated 1 message and did one more commit. Now everything is good with AppVeyor.

**Comment by XhmikosR** [comment]
How about naming the option `createEmptyArchive`? It's shorter.

**Comment by sharikovvladislav** [comment]
I was thinking about this option is too long, but didn't know how I can change it.  So yes it is possible to change the name and invert it value. I think your name is better (it is understandable and a bit shorter).

**Comment by XhmikosR** [comment]
Yeah, please do so.

File: Gruntfile.js [hunk 1]
```diff
@@ -59,6 +59,38 @@ module.exports = function(grunt) {
...
       },
+      zipDoNotCreateEmptyArchiveTrue: {
+        options: {
+          doNotCreateEmptyArchive: true,
+          archive: function () {
+            return 'tmp/compress_test_files_empty_must_not_be_created_because_option_set_to_true.zip';
+          }
+        },
+        files: [
+          {expand: true, cwd: 'test/fixtures/', src: ['NotExistingFilePath.js']}
```
Comment (XhmikosR) [inline Gruntfile.js]: Please use

```js
        files: [{
          expand: true,
          cwd: 'test/fixtures/',
          src: ['NotExistingFilePath.js']
        }]
```

Same in line 81.

File: README.md [hunk 2]
```diff
@@ -61,6 +61,14 @@ Default: `false`
 
 Pretty print file sizes when logging.
 
+#### doNotCreateEmptyArchive
+Type: `Boolean`
```
Comment (XhmikosR) [inline README.md]: Add two trailing spaces for a line break.

File: tasks/compress.js [hunk 3]
```diff
@@ -16,9 +16,15 @@ module.exports = function(grunt) {
     compress.options = this.options({
       archive: null,
       mode: null,
+      doNotCreateEmptyArchive: false,
       level: 1
     });
 
+    if (compress.options.doNotCreateEmptyArchive && this.files.length === 0) {
+      grunt.log.ok('No files found. Archive wouldn\'t be created ("doNotCreateEmptyArchive" option is set to true)');
```
Comment (XhmikosR) [inline tasks/compress.js]: **won't be created**

File: Gruntfile.js [hunk 4]
```diff
@@ -59,6 +59,38 @@ module.exports = function(grunt) {
           {expand: true, cwd: 'test/fixtures/', src: ['**'], dest: './'}
         ]
       },
+      zipDoNotCreateEmptyArchiveTrue: {
+        options: {
+          doNotCreateEmptyArchive: true,
+          archive: function () {
+            return 'tmp/compress_test_files_empty_must_not_be_created_because_option_set_to_true.zip';
```
Comment (XhmikosR) [inline Gruntfile.js]: This should be shorter; there's no gain by using such a long filename.

Comment (sharikovvladislav) [inline Gruntfile.js]: What do you suggest? I afraid that name is too long. It is so only for one reason: name explains why this file is exists. Should I use some really short name or what? Like `zip_shouldnt_be_created_1.zip`?

Comment (XhmikosR) [inline Gruntfile.js]: Yeah, that's better.

File: README.md [hunk 5]
```diff
@@ -61,6 +61,14 @@ Default: `false`
...
+Default: `false`
+
+This is used when you don't want to get empty archive as a result, when there are no files at specified paths.
+
+It may be usefull, if you don't clearly know if files are exists and you don't need empty archive as a result.
```
Comment (sharikovvladislav) [inline README.md]: Fixed

File: Gruntfile.js [hunk 6]
```diff
@@ -59,6 +59,38 @@ module.exports = function(grunt) {
...
+        },
+        files: [
+          {expand: true, cwd: 'test/fixtures/', src: ['NotExistingFilePath.js']}
+        ]
+      },
+      zipDoNotCreateEmptyArchiveFalse: {
+        options: {
+          doNotCreateEmptyArchive: false,
+          archive: function () {
+            return 'tmp/compress_test_files_empty_must_be_created_because_option_set_to_false.zip';
```
Comment (sharikovvladislav) [inline Gruntfile.js]: Fixed

File: Gruntfile.js (hunk 4 above)
Comment (sharikovvladislav) [inline Gruntfile.js]: Fixed

File: Gruntfile.js (hunk 1 above)
Comment (sharikovvladislav) [inline Gruntfile.js]: Same code on other lines in this file. Should I refactor coding style in other lines?
Lines 51, 59, 105, 113, 121, 129 in 7feb42e rev.

File: README.md (hunk 2 above)
Comment (sharikovvladislav) [inline README.md]: Fixed

**Comment by sharikovvladislav** [comment]
I renamed option to "createEmptyArchive"

File: README.md (hunk 2 above)
Comment (sharikovvladislav) [inline README.md]: Fixed

**Comment by sharikovvladislav** [comment]
I also fixed trailing spaces issue

File: Gruntfile.js [hunk 7]
```diff
@@ -59,6 +59,44 @@ module.exports = function(grunt) {
...
+        files: [{
+          expand: true,
+          cwd: 'test/fixtures/',
+          src: ['NotExistingFilePath.js']
+        }]
+      },
+      zipNoCreateEmptyArchiveOption: {
+        options: {
+          archive: function () {
+            return 'tmp/zip_must_be_created_1.zip';
```
Comment (XhmikosR) [inline Gruntfile.js]: I don't see any other file named `zip_must_be_created_1.zip` so maybe change all instances to `zip_must_be_created.zip`?

Comment (sharikovvladislav) [inline Gruntfile.js]: Oh. My bad. There must be two names for 2 cases. I will explain why.

There are 3 tests. 1nd and 3rd tests expect archive to be created (option is true or no option). If I will make same names it is possible that 1st test went good, 3rd test went bad (archive is not created) and both expects will be good, but it is not true. I think I must use 2 different names there. What do you think?

**Comment by sharikovvladislav** [comment]
Why check about CLA is not completed? I signed it about ~10 minutes ago?

**Comment by XhmikosR** [comment]
Because you keep committing with various git accounts.

**Comment by sharikovvladislav** [comment]
Hm. I have problems. 
1) I connected most of my commits to my github account, but I can't do this for 6: these comits was done with email vladislav.sharikov@gmail.com. This is not my email. It is my mistake. I can't link it with my github account
2) Anyway I have this message:
0 out of 3 committers have signed the CLA.
:x: Alina.Zaytseva
:x: Vladislav.Sharikov
:x: Sharikov Vladislav

I can't sign CLA with others emails except sharikov.vladislav@gmail.com (it is already signed).

What should I do know? Should I create new PR and commit only with correct email in my local github settings?

PS Yes I am committing from different places: my different home pc`s and work pc.

**Comment by sharikovvladislav** [comment]
Look I am trying this to replace commit author. Looks like it can help: http://stackoverflow.com/a/28845565/2898694

Also why CLA for sharikov.vladislav@gmail.com is not signed?

**Comment by sharikovvladislav** [comment]
OK. It is going to good. I will finish this in a few minutes.

**Comment by sharikovvladislav** [comment]
Ok I changed all authors (it was me, but I set up correct email): https://github.com/sharikovvladislav/grunt-contrib-compress/commits/master

Now is everything alright, correct?

**Comment by sharikovvladislav** [comment]
I made some changes according to this https://github.com/gruntjs/grunt-contrib-compress/pull/186#pullrequestreview-6430123 review.

File: test/compress_test.js [hunk 8]
```diff
@@ -7,6 +7,15 @@ var fs = require('fs');
 var unzip = require('unzip');
 var tar = require('tar');
 
+function fileExists(filePath) {
```
Comment (XhmikosR) [inline test/compress_test.js]: Any reason why you don't use `grunt.file.exists`?

Comment (sharikovvladislav) [inline test/compress_test.js]: No reasons! I fixed it!

Comment (sharikovvladislav) [inline test/compress_test.js]: I already fixed all this comments: https://github.com/gruntjs/grunt-contrib-compress/pull/186#pullrequestreview-6430123

**Comment by XhmikosR** [comment]
Looks goods but 2 final thoughts:

1. NotExisting ->NotExistent
2. Do not modify README.md; instead modify the needed file in docs

**Comment by sharikovvladislav** [comment]
1. Done 
2. Done

File: docs/compress-options.md [hunk 9]
```diff
@@ -69,6 +69,14 @@ Default: `false`
 
 Pretty print file sizes when logging.
 
+## createEmptyArchive
```
Comment (sharikovvladislav) [inline docs/compress-options.md]: I am not sure about header level. What do you think?

I ask becase all options in README.md file has h4, but in compress-options.md some has h3 and some has h2. Why so and what header is correct there?

Comment (XhmikosR) [inline docs/compress-options.md]: h2 is fine I think.

Comment (sharikovvladislav) [inline docs/compress-options.md]: Okay, nice. What about your request for changes? This one: https://github.com/gruntjs/grunt-contrib-compress/pull/186#pullrequestreview-6430123. I think I fixed all.

**Comment by XhmikosR** [comment]
@vladikoff: what do you think about this?

**Comment by sharikovvladislav** [comment]
What is the usecase? We distribute our product as zips. So our ci makes some zip files. We also store units test results and other statistics . We put it in zip files too . We have npm package that build sources in the repos. Each subsystem has this package in dependencies . This tool uses grunt-contrib-compress to compress files . Some subsystems does not have units . In that case empty zip is created . So this new option will allow everybody to not get empty archive if they don't want to .

**Comment by XhmikosR** [comment]
Something seems weird. With this patch merged, running `grunt` in this repo doesn't run the contrib tasks but only runs the test target.

**Comment by XhmikosR** [comment]
Nvm, it's not this patch to blame.

**Comment by sharikovvladislav** [comment]
Yes, I just had checked changes again. There is nothing (seems to me) that can affect contrib tasks.

**Pull request merged**